backend/junior-3/haskell.md

```diff
@@ -0,0 +1,447 @@
+  * [Wadler letter "Make ad hoc polymorphism less ad hoc"](https://homepages.inf.ed.ac.uk/wadler/papers/class-letter/class-letter.txt)
+* Final tagless
+  * [Introduction to Tagless Final](https://serokell.io/blog/2018/12/07/tagless-final)
+
+### GHC extensions
+
+* Extensions for deriving:
+  * `StandaloneDeriving`
+  * `DeriveFunctor`
+  * `DeriveFoldable`
+  * `DeriveTraversable`
+  * `GeneralizedNewtypeDeriving`
+  * `DeriveAnyClass`
+  * `DerivingStrategies`
+* `RecordWildCards`
+* `TypeOperators`
+* `FlexibleContexts` & `FlexibleInstances`
+* `MultiParamTypeClasses` & `FunctionalDependencies`
+* Type and Data Families
+  * What are the three variations of type families: closed, open, associated?
+  What are their difference and purpose?
+  * What is injectivity?
+  * Is open data family injective? Is open type family injective?
+  Would type/data family be injective if it was associated with a type class?
+  Would type/data family be injective if it would be closed or open?
+  * Why there are no closed data families?
+  * Can open type family instances overlap with parameters? Can closed typed families do it?
+  * How closed type families can help to create overlapping type classes instances?
+  * What are the differences between Functional Dependencies and associated Type Families?
+  What are the situations when the best choice would be to use associated type families, and when it would be fundeps?
+* GADTs
+  * What is the difference between GADT and ADT?
+  * What is the main goal of GADT?
+  * What is the difference between `GADTs` and phantom types?
+  * How GADT helps us to go from type level to term level in pattern matching?
+  * Are record fields available for GADT?
+  * Is it considered a good practice to put constraints in consructors inside GADT declaration?
+  * How does deriving instances work for GADT? Are there some problems?
+* Data Kinds
+  * What is the kind `*`? What is the kind `Type`?
+  * What is the data type promotion?
+  * What is the main goal for `DataKinds`?
+  * What is inconvenience we've got with DataKinds?
+  * Is apostrophe (' symbol) required in a name of a promoted data type?
+  * Are types with promoted kinds are inhabited?
+  * Could existential type be promoted?
+  * How can heterogenous lists be implemented with `DataKinds` and `GADTs`?
+  * What are the dependant types and how `DataKinds` allow us to get closer to them?
+* Polymorphic Kinds
+  * What is the main goal of `PolyKinds`?
+  * When it could be useful?
+  * What declarations would have the inferred kind `*` and not polykinded `forall k. k`?
+  * What is `Proxy` datatype? What is it used for? Why do we need `PolyKinds` to implement it?
+* `ScopedTypeVariables`
+* Higher ranked types
+  * What is a "rank" of a function?
+  * Give examples of rank-1, rank-2, rank-3 functions.
+  * In which cases do we need `ExistentialQuantification` and in which we need `RankNTypes`?
```

Comment:
`Provide example cases on when we can use 'ExistentialQuantification' and when we can use 'RankNTypes'`